app/nexus/internal/initialization/recovery/recovery.go

```diff
@@ -38,20 +39,44 @@ import (
 func RecoverBackingStoreUsingKeeperShards(source *workloadapi.X509Source) {
 	const fName = "RecoverBackingStoreUsingKeeperShards"
 
+	log.Log().Info(fName, "msg", "Recovering backing store using keeper shards")
+
 	successfulKeeperShards := make(map[string]string)
 
 	for {
 		recoverySuccessful := iterateKeepersAndTryRecovery(
 			source, successfulKeeperShards,
 		)
 		if recoverySuccessful {
+			log.Log().Info(fName, "msg", "Recovery successful")
 			return
 		}
+
+		log.Log().Warn(fName, "msg", "Recovery unsuccessful. Will retry.")
+		log.Log().Warn(fName, "msg",
+			"Successful keepers: "+string(rune(len(successfulKeeperShards))),
+		)
+		log.Log().Warn(fName, "msg", "You may need to manually bootstrap.")
+
 		log.Log().Info(fName, "msg", "Waiting for keepers to respond")
+
 		time.Sleep(5 * time.Second)
 	}
 }
 
+func RestoreBackingStoreUsingPilotShards(shards []string) {
+	firstShard := shards[0]
```

Comment:
most of these methods assume 2 out of 3 shards and those numbers are hard coded.

I'll make it configurable in follow-up commit.

Reply:
I wanted to make it work first and then work on make it "work better" and "work more flexible" :)